File: Backend (Springboot)/Annapurna/src/main/java/com/annapurna/controller/TiffinController.java
package com.annapurna.controller;

import com.annapurna.domain.Role;
import com.annapurna.domain.Tiffin;
import com.annapurna.domain.User;
import com.annapurna.dto.TiffinDTO;
import com.annapurna.repository.TiffinRepository;
import com.annapurna.security.JwtProvider;
import com.annapurna.service.TiffinService;
import com.annapurna.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Base64;
import java.util.List;

@RestController
@RequestMapping("/api/tiffin")
@RequiredArgsConstructor
public class TiffinController {

    private final TiffinService tiffinService;
    private final JwtProvider jwtProvider;
    private final UserService userService;
    private final TiffinRepository tiffinRepository;

    
    @GetMapping("/all")
    public ResponseEntity<List<TiffinDTO>> allTiffins(@RequestHeader("Authorization") String authHeader) {
        // Extract token (handles both "Bearer <token>" and raw token)
        String token = authHeader != null && authHeader.startsWith("Bearer ")
                ? authHeader.substring(7) : authHeader;
        String email = jwtProvider.getEmailFromToken(token);

        User customer = userService.getByEmail(email);
        if (customer == null) {
            throw new RuntimeException("Customer not found.");
        }
        if (customer.getRole() != Role.CUSTOMER) {
            throw new RuntimeException("Only customers can view tiffins.");
        }

        // Use pinCode field from Address entity
        String customerPincode = customer.getAddress() != null ? customer.getAddress().getPinCode() : null;

        List<Tiffin> tiffins;
        if (customerPincode == null || customerPincode.isBlank()) {
            // fallback — if customer has no pincode, return empty (safe default)
            tiffins = List.of();
        } else {
            // Use service method which uses repository query to fetch matching vendor tiffins
            tiffins = tiffinService.getTiffinsForCustomerByPinCode(customerPincode);
        }

        List<TiffinDTO> dtos = tiffins.stream().map(tiffin -> {
            TiffinDTO dto = new TiffinDTO();
            dto.setId(tiffin.getId());
            dto.setName(tiffin.getName());
            dto.setDescription(tiffin.getDescription());
            dto.setPrice(tiffin.getPrice());
            dto.setVendorName(tiffin.getVendor() != null ? tiffin.getVendor().getFullName() : null);
            dto.setCategory(tiffin.getCategory() != null ? tiffin.getCategory().name() : null);

            if (tiffin.getImage() != null) {
                String base64 = Base64.getEncoder().encodeToString(tiffin.getImage());
                dto.setImageUrl("data:image/jpeg;base64," + base64);
            } else {
                dto.setImageUrl(tiffin.getImageUrl()); // fallback to imageUrl if saved
            }
            return dto;
        }).toList();

        return ResponseEntity.ok(dtos);
    }

    
    
    @GetMapping("/all-tiffins")
    public ResponseEntity<List<TiffinDTO>> getAllTiffinsPublic() {
        List<Tiffin> tiffins = tiffinRepository.findAll();

        List<TiffinDTO> dtos = tiffins.stream().map(tiffin -> {
            TiffinDTO dto = new TiffinDTO();
            dto.setId(tiffin.getId());
            dto.setName(tiffin.getName());
            dto.setDescription(tiffin.getDescription());
            dto.setPrice(tiffin.getPrice());
            dto.setVendorName(tiffin.getVendor() != null ? tiffin.getVendor().getFullName() : null);
            dto.setCategory(tiffin.getCategory() != null ? tiffin.getCategory().name() : null);

            if (tiffin.getImage() != null) {
                String base64 = Base64.getEncoder().encodeToString(tiffin.getImage());
                dto.setImageUrl("data:image/jpeg;base64," + base64);
            } else {
                dto.setImageUrl(tiffin.getImageUrl());
            }

            return dto;
        }).toList();

        return ResponseEntity.ok(dtos);
    }

    
}

File: Backend (DNET)/Annapurna/Controllers/TiffinController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Authorization;
using Microsoft.EntityFrameworkCore;
using AnnapurnaAPI.Data;
using AnnapurnaAPI.Services;
using AnnapurnaAPI.Security;
using AnnapurnaAPI.DTOs;
using AnnapurnaAPI.Models;
using System.Security.Claims;
using System.Text;

namespace AnnapurnaAPI.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class TiffinController : ControllerBase
    {
        private readonly TiffinService _tiffinService;
        private readonly JwtService _jwtService;
        private readonly UserService _userService;
        private readonly ApplicationDbContext _context;

        public TiffinController(TiffinService tiffinService, JwtService jwtService, UserService userService, ApplicationDbContext context)
        {
            _tiffinService = tiffinService;
            _jwtService = jwtService;
            _userService = userService;
            _context = context;
        }

        [HttpGet("all")]
        [Authorize(Roles = "CUSTOMER")]
        public async Task<IActionResult> AllTiffins()
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                if (string.IsNullOrEmpty(email))
                {
                    return BadRequest(new { message = "User not found." });
                }

                var customer = await _userService.GetByEmail(email);
                if (customer == null)
                {
                    return BadRequest(new { message = "Customer not found." });
                }

                if (customer.Role != Role.CUSTOMER)
                {
                    return BadRequest(new { message = "Only customers can view tiffins." });
                }

                var customerPincode = customer.Address?.PinCode;
                List<Tiffin> tiffins;

                if (string.IsNullOrEmpty(customerPincode))
                {
                    tiffins = new List<Tiffin>();
                }
                else
                {
                    tiffins = await _tiffinService.GetTiffinsForCustomerByPinCode(customerPincode);
                }

                var dtos = tiffins.Select(tiffin => new TiffinDTO
                {
                    Id = tiffin.Id,
                    Name = tiffin.Name,
                    Description = tiffin.Description,
                    Price = tiffin.Price,
                    VendorName = tiffin.Vendor?.FullName,
                    Category = tiffin.Category.ToString(),
                    ImageUrl = tiffin.Image != null ? $"data:image/jpeg;base64,{Convert.ToBase64String(tiffin.Image)}" : tiffin.ImageUrl
                }).ToList();

                return Ok(dtos);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        [HttpGet("all-tiffins")]
        public async Task<IActionResult> GetAllTiffinsPublic()
        {
            try
            {
                var tiffins = await _context.Tiffins
                    .Include(t => t.Vendor)
                    .ToListAsync();

                var dtos = tiffins.Select(tiffin => new TiffinDTO
                {
                    Id = tiffin.Id,
                    Name = tiffin.Name,
                    Description = tiffin.Description,
                    Price = tiffin.Price,
                    VendorName = tiffin.Vendor?.FullName,
                    Category = tiffin.Category.ToString(),
                    ImageUrl = tiffin.Image != null ? $"data:image/jpeg;base64,{Convert.ToBase64String(tiffin.Image)}" : tiffin.ImageUrl
                }).ToList();

                return Ok(dtos);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }
    }
}
